File: src/scripts/check_council_layout.py
#!/usr/bin/env python3
"""CI guard for the `ai-council` skill's output-path convention.

Council artefacts (questions, responses, sessions) belong in three
canonical directories under `agents/`:

  - agents/runtime/council/questions/<topic-slug>.md       (paired with roadmap/ADR)
  - agents/runtime/council/responses/<topic-slug>.json     (paired with question)
  - agents/runtime/council/sessions/<UTC-timestamp>.json   (ad-hoc sessions)

The three canonical dirs are gitignored — the linter therefore only
catches **misplacement**, not naming-conventions inside the dirs:

  - Files at agents/ root with a council-* or .council-* prefix
    (e.g. agents/council-foo.md, agents/.council-foo.md).
  - council-* files under any other subdirectory of agents/.

`agents/evidence/audits/` is exempt — historical audit bundles are cohesive,
checked-in narratives (the canonical council dirs are gitignored)
and may legitimately include council-* artefacts as part of the
audit's evidence trail. `agents/runtime/` is exempt too — the
canonical council dirs live at `agents/runtime/council/{questions,
responses,sessions}/` and the whole `runtime/` tree is gitignored.

Failure modes are enforced by `.agent-src.uncondensed/skills/ai-council/SKILL.md`
§ "Output path convention".

Exit codes:
  0 — layout is clean.
  1 — at least one violation found; details printed to stdout.

Invocation (from project root):
  python3 scripts/check_council_layout.py
"""

from __future__ import annotations

import re
import sys
from pathlib import Path

QUIET = "--quiet" in sys.argv

AGENTS_ROOT = Path("agents")
# Canonical council dirs now live under agents/runtime/council/.
# Stored as relative POSIX paths from AGENTS_ROOT.
CANONICAL_DIRS = {
    "runtime/council/questions": ".md",
    "runtime/council/responses": ".json",
    "runtime/council/sessions": ".json",
}
# Top-level subdirectories whose contents are exempt from the layout
# check. `audits/` covers historical audit bundles. `runtime/` is the
# gitignored volatile tree (canonical council dirs live there).
EXEMPT_DIR_PREFIXES = ("audits", "runtime")
# A council artefact is a file whose name starts with `council-` or
# `.council-`. This intentionally excludes roadmaps like
# `road-to-ai-council.md` whose stem only contains the word "council".
COUNCIL_PREFIX_RE = re.compile(r"^\.?council-")


def is_council_artefact(path: Path) -> bool:
    return bool(COUNCIL_PREFIX_RE.match(path.name))


def find_violations(root: Path) -> list[str]:
    findings: list[str] = []
    if not root.is_dir():
        return findings

    # 1. Stray council artefacts at agents/ root
    for path in sorted(root.iterdir()):
        if not path.is_file():
            continue
        if is_council_artefact(path):
            findings.append(
                f"{path}: council artefact at agents/ root — move to "
                f"agents/runtime/council/questions/, agents/runtime/council/responses/, "
                f"or agents/runtime/council/sessions/ per ai-council § Output path "
                f"convention."
            )

    # 2. Council artefacts in non-canonical subdirectories
    for path in sorted(root.rglob("*")):
        if not path.is_file() or not is_council_artefact(path):
            continue
        try:
            rel = path.relative_to(root)
        except ValueError:
            continue
        if len(rel.parts) == 1:
            continue  # already handled above
        rel_posix = rel.as_posix()
        if any(rel_posix.startswith(d + "/") for d in CANONICAL_DIRS):
            continue
        if rel.parts[0].startswith(EXEMPT_DIR_PREFIXES):
            continue
        findings.append(
            f"{path}: council artefact in non-canonical directory "
            f"agents/{rel.parts[0]}/ — only "
            f"agents/runtime/council/{{questions,responses,sessions}}/ "
            f"are allowed."
        )

    return findings


def main() -> int:
    findings = find_violations(AGENTS_ROOT)
    if findings:
        print("❌  Council layout violations:\n")
        for f in findings:
            print(f"  - {f}")
        print(
            "\nRule: .agent-src.uncondensed/skills/ai-council/SKILL.md "
            '§ "Output path convention"'
        )
        return 1
    if not QUIET:
        print("✅  Council layout clean.")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: src/scripts/check_council_layout.ts
#!/usr/bin/env tsx
/**
 * CI guard for the `ai-council` skill's output-path convention.
 *
 * TypeScript twin of `src/scripts/check_council_layout.py` (ADR-089,
 * Phase 4 / Wave 4c). Mirrors the Python CLI contract EXACTLY — `--quiet`
 * flag (positional anywhere in argv), exit codes (0 clean, 1 violations),
 * stdout, byte-identical finding messages, same `agents/` scan,
 * sorted-iterdir + sorted-rglob ordering, same exemptions. No behaviour
 * changes.
 *
 * Council artefacts belong in three canonical dirs under
 * agents/runtime/council/{questions,responses,sessions}/. The linter
 * catches misplacement: stray council-* files at agents/ root, and
 * council-* files under non-canonical, non-exempt subdirectories.
 */

import * as fs from 'node:fs';
import * as path from 'node:path';
import { fileURLToPath, pathToFileURL } from 'node:url';

const _HERE = fileURLToPath(import.meta.url);

const AGENTS_ROOT = 'agents';
// Canonical council dirs (relative POSIX paths from AGENTS_ROOT).
const CANONICAL_DIRS: readonly string[] = [
    'runtime/council/questions',
    'runtime/council/responses',
    'runtime/council/sessions',
];
// Top-level subdirectories exempt from the layout check.
const EXEMPT_DIR_PREFIXES = ['audits', 'runtime'] as const;
// `council-` or `.council-` prefix (excludes `road-to-ai-council.md`).
const COUNCIL_PREFIX_RE = /^\.?council-/;

function is_council_artefact(name: string): boolean {
    return COUNCIL_PREFIX_RE.test(name);
}

function _isDir(p: string): boolean {
    try {
        return fs.statSync(p).isDirectory();
    } catch {
        return false;
    }
}

/** Mirror `sorted(root.iterdir())` — direct children, sorted by full path. */
function _iterdirSorted(root: string): string[] {
    let entries: string[];
    try {
        entries = fs.readdirSync(root);
    } catch {
        return [];
    }
    return entries.map((e) => path.join(root, e)).sort();
}

/** Mirror `sorted(root.rglob("*"))` — every descendant, sorted by full path. */
function _rglobAllSorted(root: string): string[] {
    const out: string[] = [];
    const walk = (current: string): void => {
        let entries: fs.Dirent[];
        try {
            entries = fs.readdirSync(current, { withFileTypes: true });
        } catch {
            return;
        }
        for (const entry of entries) {
            const full = path.join(current, entry.name);
            out.push(full);
            if (entry.isDirectory()) {
                walk(full);
            }
        }
    };
    walk(root);
    return out.sort();
}

function _isFile(p: string): boolean {
    try {
        return fs.statSync(p).isFile();
    } catch {
        return false;
    }
}

function find_violations(root: string): string[] {
    const findings: string[] = [];
    if (!_isDir(root)) {
        return findings;
    }

    // 1. Stray council artefacts at agents/ root.
    for (const p of _iterdirSorted(root)) {
        if (!_isFile(p)) {
            continue;
        }
        if (is_council_artefact(path.basename(p))) {
            // Mirror Python str(Path) — relative path with the AGENTS_ROOT
            // prefix as constructed (e.g. "agents/council-foo.md").
            findings.push(
                `${_posix(p)}: council artefact at agents/ root — move to ` +
                    `agents/runtime/council/questions/, agents/runtime/council/responses/, ` +
                    `or agents/runtime/council/sessions/ per ai-council § Output path ` +
                    `convention.`,
            );
        }
    }

    // 2. Council artefacts in non-canonical subdirectories.
    for (const p of _rglobAllSorted(root)) {
        if (!_isFile(p) || !is_council_artefact(path.basename(p))) {
            continue;
        }
        const rel = path.relative(root, p);
        const parts = rel.split(path.sep);
        if (parts.length === 1) {
            continue; // already handled above
        }
        const relPosix = parts.join('/');
        if (CANONICAL_DIRS.some((d) => relPosix.startsWith(d + '/'))) {
            continue;
        }
        if (EXEMPT_DIR_PREFIXES.some((d) => parts[0]!.startsWith(d))) {
            continue;
        }
        findings.push(
            `${_posix(p)}: council artefact in non-canonical directory ` +
                `agents/${parts[0]}/ — only ` +
                `agents/runtime/council/{questions,responses,sessions}/ ` +
                `are allowed.`,
        );
    }

    return findings;
}

/** Render a path as POSIX (mirrors str(Path) on the relative agents tree). */
function _posix(p: string): string {
    return p.split(path.sep).join('/');
}

function main(argv: readonly string[]): number {
    const quiet = argv.includes('--quiet');
    const findings = find_violations(AGENTS_ROOT);
    if (findings.length > 0) {
        process.stdout.write('❌  Council layout violations:\n\n');
        for (const f of findings) {
            process.stdout.write(`  - ${f}\n`);
        }
        process.stdout.write(
            '\nRule: .agent-src.uncondensed/skills/ai-council/SKILL.md ' +
                '§ "Output path convention"\n',
        );
        return 1;
    }
    if (!quiet) {
        process.stdout.write('✅  Council layout clean.\n');
    }
    return 0;
}

const _isCliEntry =
    process.argv[1] !== undefined &&
    import.meta.url === pathToFileURL(path.resolve(process.argv[1])).href;
if (_isCliEntry || process.argv[1] === _HERE) {
    process.exit(main(process.argv.slice(2)));
}

export {
    AGENTS_ROOT,
    CANONICAL_DIRS,
    EXEMPT_DIR_PREFIXES,
    COUNCIL_PREFIX_RE,
    is_council_artefact,
    find_violations,
    main,
};
